text/0072-oem.md

```diff
@@ -0,0 +1,97 @@
+# Object-Entity Mapping (OEM)
+
+## Related issues and PRs
+
+- Reference Issues: https://github.com/cedar-policy/cedar/issues/986
+- Implementation PR(s): (leave this empty)
+
+## Timeline
+
+- Started: 2024-06-28
+
+## Summary
+
+Similar to [Object-Relational Mapping (ORM)](https://en.wikipedia.org/wiki/Object%E2%80%93relational_mapping), this RFC proposes a macro-based conversion between Rust objects and entities.
+
+## Basic example
+
+The following Rust code defines how type `Team` and type `Member` map to entity type `Example::Team`.
+
+```Rust
+#[derive(Debug, Clone, Cedar)]
+#[cedar(rename="Example::Team")]
+struct Team {
+  #[cedar(eid)]
+  id: String,
+  members: HashSet<Member>
+}
+
+#[derive(Debug, Clone, Cedar)]
+#[cedar(transparent)]
+struct Member {
+  name: String,
+  #[cedar(rename="region")]
+  location: Option<String>,
+}
+```
+An object `Team { id: "A", members: { Member { name: "alice", location: Some("West")}, Member { name: "bob", location: None}}}` is converted to the following entities in JSON,
+
+```JSON
+[
+ {
+    {
+        "uid": { "type": "Team", "id": "A"},
+        "attrs": {
+            "members": [
+                {
+                    "name": "alice",
+                    "location" "West"
+                },
+                {
+                    "name": "bob"
+                }
+            ]
+        },
+        "parents": []
+    }
+ }
+]
+```
+## Motivation
+
+There currently do not exist any convenient methods to programmatically convert Rust objects to entities or vice versa. The status quo of converting a Rust object to an entity is to create an entity without any attributes and then insert attributes. The whole process is akin to that of *manually* serializing Rust objects to JSON, the latter of which is automated by serde. Users should be able to do the same for Rust object to entity conversion. This RFC addresses this need using Rust derive macros.
+
+Moreover, OEM can potentially facilitate the adoption of Cedar on applications using databases. For instance, developers can leverage ORM to seamlessly convert results of database queries into Rust objects and then into entities. That is, developers are able to conveniently construct authorization requests by fetching data from databases.
+
+## Detailed design
+
+Given limited knowledge about Rust macros the proposer has, the design details are subject to significant changes.
+
+The macros over a Rust type eventually expand to implementations of `Into<Entities>`. In the meantime, they also give rise to implementations of `Into<Value>` for this type. By default, the implementation of `Into<Value` for a type produces an entity `uid` unless the type is annotated with the `#[cedar(transparent)]` clause, where it produces a record.
```

Comment:
> produces an entity `uid` unless the type is annotated with the `#[cedar(transparent)]` clause

I would rather see this made explicit. E.g., `#[cedar(record)]` or `#[cedar(entity)]`, perhaps with the entity case being assumed by default.

This could also support a nice syntax like `#[cedar(long)] struct MyLong(i32)` for wrapper types that might exist.